File: test/testJsonBuilder.cpp
//MIT License

//Copyright (c) 2020 bexoft GmbH (mail@bexoft.de)

//Permission is hereby granted, free of charge, to any person obtaining a copy
//of this software and associated documentation files (the "Software"), to deal
//in the Software without restriction, including without limitation the rights
//to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//copies of the Software, and to permit persons to whom the Software is
//furnished to do so, subject to the following conditions:

//The above copyright notice and this permission notice shall be included in all
//copies or substantial portions of the Software.

//THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//SOFTWARE.

#include "gtest/gtest.h"
#include "gmock/gmock.h"


#include "finalmq/helpers/FmqDefines.h"
#include "finalmq/json/JsonBuilder.h"
#include "MockIZeroCopyBuffer.h"

using ::testing::_;
using ::testing::Return;
using ::testing::Invoke;
using ::testing::StrEq;
using ::testing::Sequence;

using namespace finalmq;


struct String : public std::string
{
    void resize(size_t size)
    {
        std::string::resize(size);
    }
};


class TestJsonBuilder: public testing::Test
{
public:

protected:
    virtual void SetUp()
    {
        static const int MAX_BLOCK_SIZE = 1024;

        m_data.resize(MAX_BLOCK_SIZE);
        EXPECT_CALL(m_mockBuffer, addBuffer(MAX_BLOCK_SIZE, _)).Times(1).WillOnce(Return((char*)m_data.data()));
        EXPECT_CALL(m_mockBuffer, downsizeLastBuffer(_)).Times(1).WillOnce(Invoke(&m_data, &String::resize));
        EXPECT_CALL(m_mockBuffer, getRemainingSize()).WillRepeatedly(Return(0));

        m_builder = std::unique_ptr<JsonBuilder>(new JsonBuilder(m_mockBuffer, MAX_BLOCK_SIZE));
    }

    virtual void TearDown()
    {
    }

    std::unique_ptr<IJsonParserVisitor> m_builder;
    String                              m_data;
    MockIZeroCopyBuffer                 m_mockBuffer;
};



TEST_F(TestJsonBuilder, testNull)
{
    m_builder->enterNull();
    m_builder->finished();
    EXPECT_EQ(m_data, "null");
}


TEST_F(TestJsonBuilder, testBoolTrue)
{
    static const bool VALUE = true;

    m_builder->enterBool(VALUE);
    m_builder->finished();
    EXPECT_EQ(m_data, "true");
}


TEST_F(TestJsonBuilder, testBoolFalse)
{
    static const bool VALUE = false;

    m_builder->enterBool(VALUE);
    m_builder->finished();
    EXPECT_EQ(m_data, "false");
}


TEST_F(TestJsonBuilder, testInt32_0)
{
    static const std::int32_t VALUE = 0;

    m_builder->enterInt32(VALUE);
    m_builder->finished();
    EXPECT_EQ(m_data, "0");
}

TEST_F(TestJsonBuilder, testInt32_1234567890)
{
    static const std::int32_t VALUE = 1234567890;

    m_builder->enterInt32(VALUE);
    m_builder->finished();
    EXPECT_EQ(m_data, "1234567890");
}

TEST_F(TestJsonBuilder, testInt32_n1234567890)
{
    static const std::int32_t VALUE = -1234567890;

    m_builder->enterInt32(VALUE);
    m_builder->finished();
    EXPECT_EQ(m_data, "-1234567890");
}

TEST_F(TestJsonBuilder, testUInt32_4234567890)
{
    static const std::uint32_t VALUE = 4234567890;

    m_builder->enterUInt32(VALUE);
    m_builder->finished();
    EXPECT_EQ(m_data, "4234567890");
}


TEST_F(TestJsonBuilder, testInt64_0)
{
    static const std::int64_t VALUE = 0;

    m_builder->enterInt64(VALUE);
    m_builder->finished();
    EXPECT_EQ(m_data, "0");
}

TEST_F(TestJsonBuilder, testInt64_1234567890123456789)
{
    static const std::int64_t VALUE = 1234567890123456789;

    m_builder->enterInt64(VALUE);
    m_builder->finished();
    EXPECT_EQ(m_data, "1234567890123456789");
}

TEST_F(TestJsonBuilder, testInt64_n1234567890123456789)
{
    static const std::int64_t VALUE = -1234567890123456789;

    m_builder->enterInt64(VALUE);
    m_builder->finished();
    EXPECT_EQ(m_data, "-1234567890123456789");
}

TEST_F(TestJsonBuilder, testUInt64_2234567890123456789)
{
    static const std::uint64_t VALUE = 2234567890123456789;

    m_builder->enterUInt64(VALUE);
    m_builder->finished();
    EXPECT_EQ(m_data, "2234567890123456789");
}

TEST_F(TestJsonBuilder, testDouble_22345678901234567)
{
    static const double VALUE = -0.22345678901234567;

    m_builder->enterDouble(VALUE);
    m_builder->finished();
    m_data.resize(m_data.size() - 1);   // remove last digit
    EXPECT_EQ(m_data, "-0.2234567890123456");
}

TEST_F(TestJsonBuilder, testDouble_2234567e123)
{
    static const double VALUE = 2234567e123;

    m_builder->enterDouble(VALUE);
    m_builder->finished();
    EXPECT_EQ(m_data, "2.234567e129");
}

TEST_F(TestJsonBuilder, testDouble_2234567en123)
{
    static const double VALUE = 2234567e-123;

    m_builder->enterDouble(VALUE);
    m_builder->finished();
    EXPECT_EQ(m_data, "2.234567e-117");
}

TEST_F(TestJsonBuilder, testString)
{
    static const std::string VALUE = "Hello";

    m_builder->enterString(VALUE.c_str(), VALUE.size());
    m_builder->finished();
    EXPECT_EQ(m_data, "\"Hello\"");
}

TEST_F(TestJsonBuilder, testStringEmpty)
{
    static const std::string VALUE = "";

    m_builder->enterString(VALUE.c_str(), VALUE.size());
    m_builder->finished();
    EXPECT_EQ(m_data, "\"\"");
}


TEST_F(TestJsonBuilder, testStringEscape)
{
    static const std::string VALUE = "\\\"\r\n\f\b\t\x01";

    m_builder->enterString(VALUE.c_str(), VALUE.size());
    m_builder->finished();
    EXPECT_EQ(m_data, std::string("\"\\\\\\\"\\r\\n\\f\\b\\t\\u0001\""));
}


TEST_F(TestJsonBuilder, testStringEscapeU16)
{
    static const std::string VALUE = {(char)0xc3, (char)0xa4};

    m_builder->enterString(VALUE.c_str(), VALUE.size());
    m_builder->finished();
    EXPECT_EQ(m_data, "\"\\u00e4\"");
}

TEST_F(TestJsonBuilder, testStringEscapeU32)
{
    static const std::string VALUE = {(char)0xf0, (char)0x90, (char)0xa6, (char)0xaa};

    m_builder->enterString(VALUE.c_str(), VALUE.size());
    m_builder->finished();
    EXPECT_EQ(m_data, "\"\\ud802\\uddaa\"");
}

TEST_F(TestJsonBuilder, testArray)
{
    m_builder->enterArray();
    m_builder->exitArray();
    m_builder->finished();
    EXPECT_EQ(m_data, "[]");
}

TEST_F(TestJsonBuilder, testObject)
{
    m_builder->enterObject();
    m_builder->exitObject();
    m_builder->finished();
    EXPECT_EQ(m_data, "{}");
}


TEST_F(TestJsonBuilder, testKey)
{
    static const std::string VALUE = "Hello";

    m_builder->enterKey(VALUE.c_str(), VALUE.size());
    m_builder->finished();
    EXPECT_EQ(m_data, "\"Hello\":");
}

TEST_F(TestJsonBuilder, testKeyEmpty)
{
    static const std::string VALUE = "";

    m_builder->enterKey(VALUE.c_str(), VALUE.size());
    m_builder->finished();
    EXPECT_EQ(m_data, "\"\":");
}


TEST_F(TestJsonBuilder, testKeyEscape)
{
    static const std::string VALUE = "\\\"\r\n\f\b\t\x01";

    m_builder->enterKey(VALUE.c_str(), VALUE.size());
    m_builder->finished();
    EXPECT_EQ(m_data, std::string("\"\\\\\\\"\\r\\n\\f\\b\\t\\u0001\":"));
}


TEST_F(TestJsonBuilder, testKeyEscapeU16)
{
    static const std::string VALUE = {(char)0xc3, (char)0xa4};

    m_builder->enterKey(VALUE.c_str(), VALUE.size());
    m_builder->finished();
    EXPECT_EQ(m_data, "\"\\u00e4\":");
}

TEST_F(TestJsonBuilder, testKeyEscapeU32)
{
    static const std::string VALUE = {(char)0xf0, (char)0x90, (char)0xa6, (char)0xaa};

    m_builder->enterKey(VALUE.c_str(), VALUE.size());
    m_builder->finished();
    EXPECT_EQ(m_data, "\"\\ud802\\uddaa\":");
}

TEST_F(TestJsonBuilder, testObjectExample)
{
    static const std::string KEY1 = "name";
    static const std::string NAME = "Elvis";
    static const std::string KEY2 = "age";
    static const std::string KEY3 = "arr";

    m_builder->enterObject();
    m_builder->enterKey(KEY1.c_str(), KEY1.size());
    m_builder->enterString(NAME.c_str(), NAME.size());
    m_builder->enterKey(KEY2.c_str(), KEY2.size());
    m_builder->enterInt32(42);
    m_builder->enterKey(KEY3.c_str(), KEY3.size());
    m_builder->enterArray();
    m_builder->enterDouble(1.234);
    m_builder->enterDouble(2.345);
    m_builder->enterDouble(3.456);
    m_builder->exitArray();
    m_builder->exitObject();
    m_builder->finished();
    EXPECT_EQ(m_data, "{\"name\":\"Elvis\",\"age\":42,\"arr\":[1.234,2.345,3.456]}");
}

File: csharp/test/testfinalmq/TestJsonBuilder.cs
using System;
using System.Runtime;
using System.Collections.Generic;
using System.Net;
using System.Net.Sockets;
using System.Diagnostics;

using Xunit;
using Moq;

using finalmq;
using System.Text;


namespace testfinalmq
{

    public class TestJsonBuilder : IDisposable
    {
        static readonly int MAX_BLOCK_SIZE = 1024;

        IJsonParserVisitor m_builder;
        byte[] m_data = new byte[MAX_BLOCK_SIZE];
        int m_size = MAX_BLOCK_SIZE;
        BufferRef m_bufferRef;
        Mock<IZeroCopyBuffer>? m_mockBuffer = null;

        public TestJsonBuilder()
        {
            m_mockBuffer = new Mock<IZeroCopyBuffer>();
            m_bufferRef = new BufferRef(m_data, 0, m_data.Length);

            m_mockBuffer.Setup(x => x.AddBuffer(MAX_BLOCK_SIZE, It.IsAny<int>())).Returns(m_bufferRef);
            m_mockBuffer.Setup(x => x.DownsizeLastBuffer(It.IsAny<int>())).Callback((int size) => {
                m_size = size;
            });
            m_mockBuffer.Setup(x => x.RemainingSize).Returns(0);

            m_builder = new JsonBuilder(m_mockBuffer.Object, MAX_BLOCK_SIZE);
        }

        public void Dispose()
        {
            if (m_mockBuffer != null)
            {
                m_mockBuffer.Verify(x => x.AddBuffer(MAX_BLOCK_SIZE, It.IsAny<int>()), Times.Once);
                m_mockBuffer.Verify(x => x.DownsizeLastBuffer(It.IsAny<int>()), Times.Once);
                m_mockBuffer.Verify(x => x.RemainingSize, Times.Once);
            }
        }


        [Fact]
        public void TestNull()
        {
            m_builder.EnterNull();
            m_builder.Finished();
            Debug.Assert(Encoding.UTF8.GetString(m_data, 0, m_size) == "null");
        }

        [Fact]
        public void TestTrue()
        {
            m_builder.EnterBool(true);
            m_builder.Finished();
            Debug.Assert(Encoding.UTF8.GetString(m_data, 0, m_size) == "true");
        }

        [Fact]
        public void TestFalse()
        {
            m_builder.EnterBool(false);
            m_builder.Finished();
            Debug.Assert(Encoding.UTF8.GetString(m_data, 0, m_size) == "false");
        }

        [Fact]
        public void TestInt32_0()
        {
            m_builder.EnterInt32(0);
            m_builder.Finished();
            Debug.Assert(Encoding.UTF8.GetString(m_data, 0, m_size) == "0");
        }

        [Fact]
        public void TestInt32_1234567890()
        {
            m_builder.EnterInt32(1234567890);
            m_builder.Finished();
            Debug.Assert(Encoding.UTF8.GetString(m_data, 0, m_size) == "1234567890");
        }

        [Fact]
        public void TestInt32_n1234567890()
        {
            m_builder.EnterInt32(-1234567890);
            m_builder.Finished();
            Debug.Assert(Encoding.UTF8.GetString(m_data, 0, m_size) == "-1234567890");
        }

        [Fact]
        public void TestUInt32_4234567890()
        {
            m_builder.EnterUInt32(4234567890);
            m_builder.Finished();
            Debug.Assert(Encoding.UTF8.GetString(m_data, 0, m_size) == "4234567890");
        }

        [Fact]
        public void TestInt64_0()
        {
            m_builder.EnterInt64(0);
            m_builder.Finished();
            Debug.Assert(Encoding.UTF8.GetString(m_data, 0, m_size) == "0");
        }

        [Fact]
        public void TestInt64_1234567890123456789()
        {
            m_builder.EnterInt64(1234567890123456789);
            m_builder.Finished();
            Debug.Assert(Encoding.UTF8.GetString(m_data, 0, m_size) == "1234567890123456789");
        }

        [Fact]
        public void TestInt64_n1234567890123456789()
        {
            m_builder.EnterInt64(-1234567890123456789);
            m_builder.Finished();
            Debug.Assert(Encoding.UTF8.GetString(m_data, 0, m_size) == "-1234567890123456789");
        }

        [Fact]
        public void TestUInt64_2234567890123456789()
        {
            m_builder.EnterUInt64(2234567890123456789);
            m_builder.Finished();
            Debug.Assert(Encoding.UTF8.GetString(m_data, 0, m_size) == "2234567890123456789");
        }

        [Fact]
        public void TestDouble_22345678901234567()
        {
            m_builder.EnterDouble(-0.22345678901234567);
            m_builder.Finished();
            string s = Encoding.UTF8.GetString(m_data, 0, m_size - 1);
            Debug.Assert(s == "-0.2234567890123456");
        }

        [Fact]
        public void TestDouble_2234567en123()
        {
            m_builder.EnterDouble(2234567e-123);
            m_builder.Finished();
            string s = Encoding.UTF8.GetString(m_data, 0, m_size);
            Debug.Assert(s == "2.234567E-117");
        }

        [Fact]
        public void TestString()
        {
            m_builder.EnterString("Hello");
            m_builder.Finished();
            string s = Encoding.UTF8.GetString(m_data, 0, m_size);
            Debug.Assert(s == "\"Hello\"");
        }

        [Fact]
        public void TestStringEmpty()
        {
            m_builder.EnterString("");
            m_builder.Finished();
            string s = Encoding.UTF8.GetString(m_data, 0, m_size);
            Debug.Assert(s == "\"\"");
        }

        [Fact]
        public void TestStringEscape()
        {
            m_builder.EnterString("\\\"\r\n\f\b\t\x01");
            m_builder.Finished();
            string s = Encoding.UTF8.GetString(m_data, 0, m_size);
            Debug.Assert(s == "\"\\\\\\\"\\r\\n\\f\\b\\t\\u0001\"");
        }

        [Fact]
        public void TestStringEscapeU16()
        {
            m_builder.EnterString(Encoding.UTF8.GetString(new byte[] { 0xc3, 0xa4 }));
            m_builder.Finished();
            string s = Encoding.UTF8.GetString(m_data, 0, m_size);
            Debug.Assert(s == "\"\\u00e4\"");
        }

        [Fact]
        public void TestStringEscapeU32()
        {
            m_builder.EnterString(Encoding.UTF8.GetString(new byte[] { 0xf0, 0x90, 0xa6, 0xaa }));
            m_builder.Finished();
            string s = Encoding.UTF8.GetString(m_data, 0, m_size);
            Debug.Assert(s == "\"\\ud802\\uddaa\"");
        }

        [Fact]
        public void TestStringBuffer()
        {
            m_builder.EnterString(Encoding.UTF8.GetBytes("Hello"), 0, 5);
            m_builder.Finished();
            string s = Encoding.UTF8.GetString(m_data, 0, m_size);
            Debug.Assert(s == "\"Hello\"");
        }

        [Fact]
        public void TestStringBufferEmpty()
        {
            m_builder.EnterString(Encoding.UTF8.GetBytes(""), 0, 0);
            m_builder.Finished();
            string s = Encoding.UTF8.GetString(m_data, 0, m_size);
            Debug.Assert(s == "\"\"");
        }

        [Fact]
        public void TestStringBufferEscape()
        {
            m_builder.EnterString(Encoding.UTF8.GetBytes("\\\"\r\n\f\b\t\x01"), 0, 8);
            m_builder.Finished();
            string s = Encoding.UTF8.GetString(m_data, 0, m_size);
            Debug.Assert(s == "\"\\\\\\\"\\r\\n\\f\\b\\t\\u0001\"");
        }

        [Fact]
        public void TestStringBufferEscapeU16()
        {
            m_builder.EnterString(new byte[] { 0xc3, 0xa4 }, 0, 2);
            m_builder.Finished();
            string s = Encoding.UTF8.GetString(m_data, 0, m_size);
            Debug.Assert(s == "\"\\u00e4\"");
        }

        [Fact]
        public void TestStringBufferEscapeU32()
        {
            m_builder.EnterString(new byte[] { 0xf0, 0x90, 0xa6, 0xaa }, 0, 4);
            m_builder.Finished();
            string s = Encoding.UTF8.GetString(m_data, 0, m_size);
            Debug.Assert(s == "\"\\ud802\\uddaa\"");
        }

        [Fact]
        public void TestArray()
        {
            m_builder.EnterArray();
            m_builder.ExitArray();
            m_builder.Finished();
            string s = Encoding.UTF8.GetString(m_data, 0, m_size);
            Debug.Assert(s == "[]");
        }

        [Fact]
        public void TestObject()
        {
            m_builder.EnterObject();
            m_builder.ExitObject();
            m_builder.Finished();
            string s = Encoding.UTF8.GetString(m_data, 0, m_size);
            Debug.Assert(s == "{}");
        }
        
        [Fact]
        public void TestKey()
        {
            m_builder.EnterKey("Hello");
            m_builder.Finished();
            string s = Encoding.UTF8.GetString(m_data, 0, m_size);
            Debug.Assert(s == "\"Hello\":");
        }

        [Fact]
        public void TestKeyEmpty()
        {
            m_builder.EnterKey("");
            m_builder.Finished();
            string s = Encoding.UTF8.GetString(m_data, 0, m_size);
            Debug.Assert(s == "\"\":");
        }

        [Fact]
        public void TestKeyEscape()
        {
            m_builder.EnterKey("\\\"\r\n\f\b\t\x01");
            m_builder.Finished();
            string s = Encoding.UTF8.GetString(m_data, 0, m_size);
            Debug.Assert(s == "\"\\\\\\\"\\r\\n\\f\\b\\t\\u0001\":");
        }

        [Fact]
        public void TestKeyEscapeU16()
        {
            m_builder.EnterKey(Encoding.UTF8.GetString(new byte[] { 0xc3, 0xa4 }));
            m_builder.Finished();
            string s = Encoding.UTF8.GetString(m_data, 0, m_size);
            Debug.Assert(s == "\"\\u00e4\":");
        }

        [Fact]
        public void TestKeyEscapeU32()
        {
            m_builder.EnterKey(Encoding.UTF8.GetString(new byte[] { 0xf0, 0x90, 0xa6, 0xaa }));
            m_builder.Finished();
            string s = Encoding.UTF8.GetString(m_data, 0, m_size);
            Debug.Assert(s == "\"\\ud802\\uddaa\":");
        }

        [Fact]
        public void TestKeyBuffer()
        {
            m_builder.EnterKey(Encoding.UTF8.GetBytes("Hello"), 0, 5);
            m_builder.Finished();
            string s = Encoding.UTF8.GetString(m_data, 0, m_size);
            Debug.Assert(s == "\"Hello\":");
        }

        [Fact]
        public void TestKeyBufferEmpty()
        {
            m_builder.EnterKey(Encoding.UTF8.GetBytes(""), 0, 0);
            m_builder.Finished();
            string s = Encoding.UTF8.GetString(m_data, 0, m_size);
            Debug.Assert(s == "\"\":");
        }

        [Fact]
        public void TestKeyBufferEscape()
        {
            m_builder.EnterKey(Encoding.UTF8.GetBytes("\\\"\r\n\f\b\t\x01"), 0, 8);
            m_builder.Finished();
            string s = Encoding.UTF8.GetString(m_data, 0, m_size);
            Debug.Assert(s == "\"\\\\\\\"\\r\\n\\f\\b\\t\\u0001\":");
        }

        [Fact]
        public void TestKeyBufferEscapeU16()
        {
            m_builder.EnterKey(new byte[] { 0xc3, 0xa4 }, 0, 2);
            m_builder.Finished();
            string s = Encoding.UTF8.GetString(m_data, 0, m_size);
            Debug.Assert(s == "\"\\u00e4\":");
        }

        [Fact]
        public void TestKeyBufferEscapeU32()
        {
            m_builder.EnterKey(new byte[] { 0xf0, 0x90, 0xa6, 0xaa }, 0, 4);
            m_builder.Finished();
            string s = Encoding.UTF8.GetString(m_data, 0, m_size);
            Debug.Assert(s == "\"\\ud802\\uddaa\":");
        }

        [Fact]
        public void TestObjectExample()
        {
            m_builder.EnterObject();
            m_builder.EnterKey("name");
            m_builder.EnterString("Elvis");
            m_builder.EnterKey("age");
            m_builder.EnterInt32(42);
            m_builder.EnterKey("arr");
            m_builder.EnterArray();
            m_builder.EnterDouble(1.234);
            m_builder.EnterDouble(2.345);
            m_builder.EnterDouble(3.456);
            m_builder.ExitArray();
            m_builder.ExitObject();
            m_builder.Finished();

            string s = Encoding.UTF8.GetString(m_data, 0, m_size);
            Debug.Assert(s == "{\"name\":\"Elvis\",\"age\":42,\"arr\":[1.234,2.345,3.456]}");
        }

        [Fact]
        public void TestWithRealZeroCopyBuffer()
        {
            m_mockBuffer = null;

            IMessage message = new ProtocolMessage(0);
            m_builder = new JsonBuilder(message, 8);

            m_builder.EnterObject();
            m_builder.EnterKey("name");
            m_builder.EnterString("This is Elvis");
            m_builder.EnterKey("age");
            m_builder.EnterInt32(42);
            m_builder.EnterKey("arr");
            m_builder.EnterArray();
            m_builder.EnterString("This is a loooooooooooooooooooooooooooooooooooooooooooong text");
            m_builder.EnterDouble(1.234);
            m_builder.EnterDouble(2.345);
            m_builder.EnterDouble(3.456);
            m_builder.ExitArray();
            m_builder.ExitObject();
            m_builder.Finished();

            string result = "";
            IList<BufferRef> buffers = message.GetAllSendBuffers();
            foreach (var buffer in buffers)
            {
                result += Encoding.UTF8.GetString(buffer.Buffer, buffer.Offset, buffer.Length);
            }
            Debug.Assert(result == "{\"name\":\"This is Elvis\",\"age\":42,\"arr\":[\"This is a loooooooooooooooooooooooooooooooooooooooooooong text\",1.234,2.345,3.456]}");
        }

        [Fact]
        public void TestWithRealZeroCopyBufferPreAllocateBuffer()
        {
            m_mockBuffer = null;

            IMessage message = new ProtocolMessage(0);
            BufferRef preBuffer = message.AddBuffer(40, 100);
            Encoding.UTF8.GetBytes("PRE").CopyTo(preBuffer.Buffer, preBuffer.Offset);
            message.DownsizeLastBuffer(3);

            m_builder = new JsonBuilder(message, 8);

            m_builder.EnterObject();
            m_builder.EnterKey("name");
            m_builder.EnterString("This is Elvis");
            m_builder.EnterKey("age");
            m_builder.EnterInt32(42);
            m_builder.EnterKey("arr");
            m_builder.EnterArray();
            m_builder.EnterString("This is a loooooooooooooooooooooooooooooooooooooooooooong text");
            m_builder.EnterDouble(1.234);
            m_builder.EnterDouble(2.345);
            m_builder.EnterDouble(3.456);
            m_builder.ExitArray();
            m_builder.ExitObject();
            m_builder.Finished();

            string result = "";
            IList<BufferRef> buffers = message.GetAllSendBuffers();
            foreach (var buffer in buffers)
            {
                result += Encoding.UTF8.GetString(buffer.Buffer, buffer.Offset, buffer.Length);
            }
            Debug.Assert(result == "PRE{\"name\":\"This is Elvis\",\"age\":42,\"arr\":[\"This is a loooooooooooooooooooooooooooooooooooooooooooong text\",1.234,2.345,3.456]}");
        }
    }
}

